edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -0,0 +1,423 @@
+                new Exception("Could not store message id to monitor Mqtt ACK");
+            }
+
+            var result = await tcs.Task;
+
+            return result;
+        }
+
+        void ForwardPublish(object sender, MqttMsgPublishEventArgs e)
+        {
+            var isWritten = this.publications.Match(
+                                    channel => channel.Writer.TryWrite(new MqttPublishInfo(e.Topic, e.Message)),
+                                    () => false);
+
+            if (!isWritten)
+            {
+                // Dropping the message
+                Events.CouldNotForwardMessage(e.Topic, e.Message.Length);
+            }
+        }
+
+        void ConfirmPublished(object sender, MqttMsgPublishedEventArgs e)
+        {
+            lock (this.guard)
+            {
+                if (this.pendingAcks.TryRemove(e.MessageId, out TaskCompletionSource<bool> tcs))
+                {
+                    tcs.SetResult(e.IsPublished);
+                }
+                else
+                {
+                    Events.UnknownMessageId(e.MessageId);
+                }
+            }
+        }
+
+        void TriggerReconnect(object sender, EventArgs e)
+        {
+            Task.Run(async () =>
+            {
+                var client = default(MqttClient);
+
+                lock (this.guard)
+                {
+                    if (!this.mqttClient.HasValue)
+                    {
+                        return;
+                    }
+
+                    client = this.mqttClient.Expect(() => new Exception("No mqtt-bridge connector instance found to use"));
+                }
+
+                // don't trigger it to ourselves while we are trying
+                client.ConnectionClosed -= this.TriggerReconnect;
+
+                var isConnected = false;
+
+                while (!isConnected)
+                {
+                    await Task.Delay(ReconnectDelayMs);
+
+                    lock (this.guard)
+                    {
+                        // seems Disconnect has been called since.
+                        if (!this.mqttClient.HasValue)
+                        {
+                            return;
+                        }
+
+                        client = this.mqttClient.Expect(() => new Exception("No mqtt-bridge connector instance found to use"));
+                    }
+
+                    isConnected = await TryConnectAsync(client, this.components.Consumers, this.systemComponentIdProvider.EdgeHubBridgeId);
+                }
+
+                client.ConnectionClosed += this.TriggerReconnect;
+            });
+        }
+
+        Task StartForwardingLoop()
+        {
+            var loopTask = Task.Run(
+                                async () =>
+                                {
+                                    Events.ForwardingLoopStarted();
+                                    while (await this.publications.Expect(ChannelIsBroken).Reader.WaitToReadAsync())
```

Comment:
I would avoid making publications an option and init it in the ctor.. 
but if it is an option, then take it into a local variable to avoid doing multiple Expect calls..